tox.ini

```diff
@@ -0,0 +1,80 @@
+[tox]
+minversion = 2.0
+envlist = py{27,35,36,37}-{test,integration,benchmark}, sdist, flake8
+
+[pytest]
+addopts = -rfxEXs --durations=20 --showlocals --reruns 3 --reruns-delay 1
+
+[flake8]
+ignore = E12, W503, E226
+max-line-length = 110
+show-source = True
+
+
+[testenv]
+recreate = True
+whitelist_externals =
+    sh
+    bash
+
+deps =
+    .[test]
+
+    integration: numpy
+
+    benchmark: pytest_benchmark
+    benchmark: awscli
+
+setenv =
+    SO_DISABLE_MOCKS={env:SO_DISABLE_MOCKS:}
+    SO_DISABLE_MOTO_SERVER={env:SO_DISABLE_MOTO_SERVER:}
+    SO_S3_URL={env:SO_S3_URL:}
+    SO_S3_RESULT_URL={env:SO_S3_RESULT_URL:}
+    AWS_ACCESS_KEY_ID={env:AWS_ACCESS_KEY_ID:}
+    AWS_SECRET_ACCESS_KEY={env:AWS_SECRET_ACCESS_KEY:}
+    TRAVIS_SECURE_ENV_VARS={env:TRAVIS_SECURE_ENV_VARS:}
+    RUN_BENCHMARKS={env:RUN_BENCHMARKS:}
+
+commands =
+    test: pytest smart_open/tests/
+
+    integration: pytest integration-tests/test_http.py integration-tests/test_207.py
+    integration: bash -c 'if [[ "$TRAVIS_SECURE_ENV_VARS" = "true" ]]; then pytest integration-tests/test_s3_ported.py; fi'
+
+    benchmark: bash -c 'if [[ "$TRAVIS_SECURE_ENV_VARS" = "true" && "$RUN_BENCHMARKS" = "true" ]]; then export SO_S3_URL=$SO_S3_URL/$(python -c "from uuid import uuid4;print(uuid4())"); set -e; pytest integration-tests/test_s3.py --benchmark-save=`git rev-parse HEAD`; set +e; aws s3 cp .benchmarks/*/*.json $SO_S3_RESULT_URL; aws s3 rm --recursive $SO_S3_URL; fi'
```

Comment:
Is there any way to make this prettier? Having it as a single long line is pain.

You could move out the git-rev-parse bit to a separate environment variable. I have a suspicion travis may already expose that for you.

The same thing goes for generating SO_S3_URL.

Reply:
Also, can we move the conditionals somewhere else?

Reply:
> Is there any way to make this prettier? Having it as a single long line is pain.

`tox ` have issues with multiline *sh commands, sorry :( I already tried to make a multi-line version, but this doesn't work

> You could move out the git-rev-parse bit to a separate environment variable. I have a suspicion travis may already expose that for you.

Yes, I can, but in that case that will work only in travis (I can't fallback to `git rev-parse` in `setenv` section)

> Also, can we move the conditionals somewhere else?

Why?

Reply:
It just looks a bit ugly and difficult to maintain.

How about making a separate bash script `ci/run_benchmarks.sh`? You could invoke it directly with `bash ci/run_benchmarks.sh` and it would read environment variables, etc. and do the hard work.

Similarly, with test_s3_ported, we could move the conditional into that script. Check that the creds are available in os.environ, and skip the tests if they are not.

Reply:
I see 3 variants here:

Note: by "check env" I mean `if [[ ... = "true" && ]] then ... fi;` construction

1. Check env directly in `sh` script and call them from tox
2. Chek env in `tox.ini` and call `.sh`
3. Check env in `.travis` file and call .sh`

what are you prefer? I dislike (3), but (1) and (2) look fine for me.

Reply:
I prefer 1).